crates/cheatnet/src/runtime_extensions/observer_extension.rs

```diff
@@ -0,0 +1,63 @@
+use std::collections::HashMap;
+
+use anyhow::{Context, Error};
+use cairo_felt::Felt252;
+use cairo_lang_runner::short_string::as_cairo_short_string;
+
+use runtime::{utils::BufferReader, ExtendedRuntime, ExtensionLogic, StarknetRuntime};
+use scarb_api::StarknetContractArtifacts;
+use starknet_api::core::ClassHash;
+
+use super::forge_runtime_extension::cheatcodes::declare::get_class_hash;
+
+#[derive(Default)]
+pub struct ObserverState {
+    pub contracts: HashMap<String, StarknetContractArtifacts>,
+    pub contract_name_to_class_hash: HashMap<String, ClassHash>,
+}
+
+pub struct ObserverExtension<'a> {
+    pub observer_state: &'a mut ObserverState,
+}
+
+impl<'a> ObserverExtension<'a> {
+    pub fn from(observer_state: &'a mut ObserverState) -> Self {
+        ObserverExtension { observer_state }
+    }
+}
+
+pub type ObserverRuntime<'a> = ExtendedRuntime<ObserverExtension<'a>>;
+
+impl<'a> ExtensionLogic for ObserverExtension<'a> {
+    type Runtime = StarknetRuntime<'a>;
+
+    fn handle_cheatcode_signal(
+        &mut self,
+        selector: &str,
+        inputs: Vec<Felt252>,
+        _extended_runtime: &mut Self::Runtime,
+    ) {
+        if selector == "declare" {
```

Comment:
If we don't want to add address -> name learning then we need to adjust the creation of `CallEntryPoint` in the entire codebase so `class_hash` field is always `Some(class_hash)` (currently it is `None` if it can be deduced from `contract_address` like comment in blockifier code suggests:

https://github.com/starkware-libs/blockifier/blob/a4c66192923552bb6f9c72e2f2a83a60c325fbfc/crates/blockifier/src/execution/entry_point.rs#L47